### PR TITLE
metrics: etcd_object_counts->apiserver_storage_object_counts(stable)

### LINKED ISSUE #98270: Create a stable version of etcd_object_counts
**What would you like to be added**:

We want to start migrating metrics which people depend on for operations to stable status. etcd_object_counts was mentioned in a recent SIG meeting as a viable candidate. However, the 'etcd_' prefix is not in line with stated metric conventions. SIG consensus determined that creating a new version of the metric with the metric name 'apiserver_storage_object_counts' would be sufficient allow this metric to be promoted. We'd want to simultaneously deprecate 'etcd_object_counts' (since it would be a duplicate) and remove it after the deprecation grace period window. 

/assign @erain 
/cc @YoyinZyc

### PR DESCRIPTION
**Pull request by erain**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

https://github.com/kubernetes/kubernetes/issues/98270#issue-791357977

#### Which issue(s) this PR fixes:

Fixes #98270

#### Does this PR introduce a user-facing change?

```release-note
Rename metrics `etcd_object_counts` to `apiserver_storage_object_counts` and mark it as stable. The original `etcd_object_counts` metrics name is marked as "Deprecated" and will be removed in the future. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1209-metrics-stability

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 1]
```diff
@@ -43,14 +43,22 @@ var (
 		},
 		[]string{"operation", "type"},
 	)
-	objectCounts = compbasemetrics.NewGaugeVec(
+	etcdObjectCounts = compbasemetrics.NewGaugeVec(
 		&compbasemetrics.GaugeOpts{
 			Name:           "etcd_object_counts",
-			Help:           "Number of stored objects at the time of last check split by kind.",
+			Help:           "(Deprecated) Number of stored objects at the time of last check split by kind. This metric is replaced by apiserver_storage_object_counts.",
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:49]: Oh you don't actually need to include the 'Deprecated' part in the help text. 
You can use: `				DeprecatedVersion: "1.21.0",` in the GaugeOpts directly.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 2]
```diff
@@ -77,7 +85,7 @@ var (
 		[]string{},
 	)
 )
-
+etcdObjectCounts
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:88]: is this intentional?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 3]
```diff
@@ -86,15 +94,17 @@ func Register() {
 	registerMetrics.Do(func() {
 		legacyregistry.MustRegister(etcdRequestLatency)
 		legacyregistry.MustRegister(objectCounts)
-		legacyregistry.MustRegister(dbTotalSize)
+    legacyregistry.MustRegister(etcdObjectCounts)
+    etcdObjectCountstry.MustRegister(dbTotalSize)
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:98]: you may need to run gofmt and golint.

IIRC: 
```shell
./hack/update_gofmt.sh
./hack/update_golint.sh
```

Comment (erain) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:98]: Done

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go (hunk 2 above)
Comment (erain) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go]: A typo. Removed.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go (hunk 1 above)
Comment (erain) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go]: Done

**Comment by erain** [comment]
Thanks for the review @logicalhan ! All comments addressed.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go [hunk 4]
```diff
@@ -86,15 +95,17 @@ func Register() {
 	registerMetrics.Do(func() {
 		legacyregistry.MustRegister(etcdRequestLatency)
 		legacyregistry.MustRegister(objectCounts)
-		legacyregistry.MustRegister(dbTotalSize)
+		legacyregistry.MustRegister(etcdObjectCounts)
+		etcdObjectCountstry.MustRegister(dbTotalSize)
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:99]: typo: should be etcdObjectCounts

Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:99]: ou actually this should be legacyregistry

Comment (erain) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/metrics/metrics.go:99]: Fixed and thanks! 

I just noticed the same thing from the failing tests... No idea why it got in.

**Comment by logicalhan** [comment]
@brancz wants this to be named `apiserver_storage_objects` but actually I am okay with `apiserver_storage_object_counts` as a convention for gauge counters. 

/hold
(wait on @brancz's reply)

**Comment by logicalhan** [comment]
@erain, can you amend the commit message? It seems that `#` is forbidden in commit messages.

**Comment by erain** [comment]
> @erain, can you amend the commit message? It seems that `#` is forbidden in commit messages.

Thanks. I've updated the commit message and also rebased on master. (It seems another stable metrics got merged before this one :D)

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve
/hold

**Review by logicalhan — APPROVED** [review]
/lgtm

**Pull request merged**